pkg/entrypoint/entrypointer.go

```diff
@@ -336,3 +343,153 @@ func (e Entrypointer) waitingCancellation(ctx context.Context, cancel context.Ca
+
+// getStepResultPath gets the path to the step result
+func getStepResultPath(stepDir string, stepName string, resultName string) string {
+	return filepath.Join(stepDir, stepName, "results", resultName)
+}
+
+// findReplacement looks for any usage of step results in an input string.
+// If found, it loads the results from the previous steps and provides the replacement value.
+func findReplacement(stepDir string, s string) (string, []string, error) {
+	value := strings.TrimSuffix(strings.TrimPrefix(s, "$("), ")")
+	stepName, resultName, arrayIdx, objectKey, paramType, err := v1.ParseStepExpression(value)
+	if err != nil {
+		return "", nil, err
+	}
+	result, err := loadStepResult(stepDir, stepName, resultName)
+	if err != nil {
+		return "", nil, err
+	}
+	replaceWithArray := []string{}
+	replaceWithString := ""
+
+	switch paramType {
+	case v1.ParamTypeObject:
+		if objectKey != "" {
+			replaceWithString = result.ObjectVal[objectKey]
+		}
+	case v1.ParamTypeArray:
+		if arrayIdx != nil {
+			replaceWithString = result.ArrayVal[*arrayIdx]
+		} else {
+			replaceWithArray = append(replaceWithArray, result.ArrayVal...)
+		}
+	// ParamTypeString
+	default:
+		replaceWithString = result.StringVal
+	}
+	return replaceWithString, replaceWithArray, nil
+}
+
+// replaceCommandAndArgs performs replacements for step results in environment variables.
+func replaceEnv(stepDir string, stepResultRegex *regexp.Regexp) error {
+	for _, e := range os.Environ() {
+		pair := strings.SplitN(e, "=", 2)
+		matches := stepResultRegex.FindAllStringSubmatch(pair[1], -1)
+		for _, m := range matches {
+			replaceWith, _, err := findReplacement(stepDir, m[0])
+			if err != nil {
+				return err
+			}
+			os.Setenv(pair[0], strings.ReplaceAll(pair[1], m[0], replaceWith))
+		}
+	}
+	return nil
+}
+
+// replaceScript performs replacements for step results in the script file.
+// It reads the original script file, performs the replacements and writes a new file
+// since the original script location is unknown.
+// It then updates the command to execute the new file.
+func replaceScript(e *Entrypointer, stepDir string, scriptPrefix string, newFilePath string, stepResultRegex *regexp.Regexp) error {
+	if len(e.Command) > 0 && strings.HasPrefix(e.Command[0], scriptPrefix) {
+		fileContentBytes, err := os.ReadFile(e.Command[0])
+		if err != nil {
+			return err
+		}
+		fileContents := string(fileContentBytes)
+		matches := stepResultRegex.FindAllStringSubmatch(fileContents, -1)
+		if len(matches) > 0 {
+			for _, m := range matches {
+				replaceWith, _, err := findReplacement(stepDir, m[0])
+				if err != nil {
+					return err
+				}
+				fileContents = strings.ReplaceAll(fileContents, m[0], replaceWith)
+			}
+			// copy the modified contents to a different file.
+			err := os.WriteFile(newFilePath, []byte(fileContents), 0600)
+			if err != nil {
+				return err
+			}
+			// set the command to execute the new file.
+			e.Command[0] = newFilePath
+		}
+	}
+	return nil
+}
```

Comment:
Allowing arbitrary replacement in scripts feels **very** sketchy to me, because of the risk of shell injection attacks. I'd prefer to guide people to go through environment variables for script usage.

Reply:
Agreed. I can call it out in the docs where we showcase this feature that they should go through the environment variables instead of scripts to avoid shell injection attacks. Isn't the same fear also true for normal substitutions?

Reply:
I added the CAUTION in the notes.

Reply:
> Isn't the same fear also true for normal substitutions?

Yes 😢 

If we were to have [Scorecards integration with Tekton, we'd probably call this out as a dangerous workflow](https://github.com/ossf/scorecard/blob/main/docs/checks.md#dangerous-workflow) that people probably shouldn't use for the same reasons listed for [GitHub Actions](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#understanding-the-risk-of-script-injections).

I think the main reason we don't remove it is v1 compatibility - I don't think we should add similar behavior for actions, even if it means deviating from current Step behavior.

Reply:
I think the issue is that we pipe these results into step actions via params and we allow that param to be used anywhere `env, command+args, and scripts`. 

```
kind: StepAction
spec:
  params:
    - name: foo
  script:
    echo $(params.foo)
```

```
kind: Task
spec:
  steps:
    - name: step1
      results:
        - name: result1
      script: |
        echo hi >> $(step.results.result1.path)
    - name: step2
      ref:
        name: stepaction
      params:
        - name: foo
           value: $(step.step1.results.result1)
```

So we would need to disallow any substitutions in `script` when it comes to `step actions`

Reply:
I'm very comfortable making things more secure by default for `StepActions`. For example, if it makes sense to not allow `param` substitutions in scripts inside `StepActions`, that is ok. 

My guess users migrating from inlined steps to `StepActions` will have to do this conversion and in turn, make things more secure by default. 

However, I think this is a topic we should discuss separately in the new year while `StepActions` is still alpha. I've created an [issue](https://github.com/tektoncd/pipeline/issues/7497) to capture this and also added it to the list in https://github.com/tektoncd/pipeline/issues/7424.

Reply:
@wlynch would you like to provide some context in the issue https://github.com/tektoncd/pipeline/issues/7497 that is tracking this?

Reply:
/hold converge on discussion in https://github.com/tektoncd/pipeline/issues/7497

Reply:
/hold cancel

Reply:
Removing step result substitutions in script.